### ISSUE TITLE
Spawn sequence cannot be completed if the player dies before spawning

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#3513]
### Issue description
In PM4, the only time a player can complete the spawn sequence is when the `PreSpawnPacketHandler` is set. This is a problem because this handler is replaced if the player dies with a `DeathPacketHandler`, which will cause the player to follow normal respawn routine but will never fire the first-join routine (e.g. firing PlayerJoinEvent, making the player not be immobile)

### Steps to reproduce the issue
1. Stand next to a cactus so you're getting damaged
2. Quit the server
3. Join the server
4. notice you can't move and there's no "joined the game" message

### OS and versions

* PocketMine-MP: b09379151a9ca5e9195750cc479ec06f46213537
* PHP: 7.3.16
* Server OS: Windows
* Game version: Win10

**Comment by dktapps** [comment]
It appears this bug occurs in the time between `NetworkSession->onTerrainReady()` and `Player->doFirstSpawn()` being called. Since `Player->spawned` becomes `true` directly before `NetworkSession->onTerrainReady()`, the player becomes "active", but the "first join" sequence doesn't execute until the response `SetLocalPlayerAsInitializedPacket` is received.

TL;DR: this is an async handling bug.

**Comment by dktapps** [comment]
This bug appears to have been introduced to PM4 by https://github.com/pmmp/PocketMine-MP/commit/939dfd9269df0feaff5b96d0dd628055b9706b95, which for some reason started to use the `spawned` flag directly after sending the final needed chunk for spawn. There's a TODO which states "not sure this should be here", which disappeared in a later refactor and thus never got the proper attention it needed.
